remove the old runtime

zerbina opened the pull request:
## Summary

Remove most of the old runtime, which became obsolete with the
removal of the old GCs. This significantly shrinks downs the `system`
module and its includes, making it easier to maintain and further
refactor them.

Public interface procedures and types related to the old GCs are also
removed. These are:
- the `ForeignCell` API
- `GC_ref` and `GC_unref` for `seq`s and `string`s
- the `GC_setStrategy` procedure

In addition, tracking both implicit and explicit `alloc` and `free`
(when using `--memTracker:on`) now works again.

## Details

Regarding the GC, the removed ones are `boehm`, `go`, `markAndSweep`,
`refc`, `regions`, `gc2`, `hooks`, and `none`. Their direct
implementation is removed, as well as all usages of the conditional
symbols related to them, which are: `boehmgc`, `gogc`, `gcMarkAndSweep`,
`gcRegions`, `gcHooks`, and `nogc`.

Without any stack-scanning GCs, the allocator doesn't need the
`isAllocatedPtr` routine anymore, so it is removed. There were also some
optional checks and `track` calls that were not enabled when using the
new runtime; they're ported over.

A list of removals and changes related to the old
runtime follows:
* remove the version 1 `seq` and `string` implementation
* remove the version 1 `repr` implementation
* keep the old RTTI (`TNimType`; required for `deepcopy` and
  `typeinfo`), but remove the now obsolete fields (`marker` and
  `finalizer`)
* remove `objectInit`
* keep the parts of `genericAssign` that the `typeinfo` module requires
* keep `genericReset`; the deepcopy and `genericAssign` implementation
  need it
* remove various other `.compilerprocs` related to the old runtime

Usage of the `nimV2`, `nimSeqsV2`, and `gcDestructors` conditional
symbol in code that is only available with the C backend are also
removed, as the symbols are always present there.

Finally, the now obsolete internal `cellseqs_v1` module is removed.

zerbina committed:
runtime: partially remove reprv1

The version for the JavaScript target is kept for now.

zerbina committed:
threads: remove old-runtime support

zerbina committed:
typeinfo: remove old-runtime support

zerbina committed:
runtime: remove the seqsv1 implementation

zerbina committed:
runtime: remove the legacy GCs

* remove their integration in `mmdisp`
* remove all other integration in `system` and `threads`
* remove the actual implementation

Keep a trimmed down `gc_common.nim` for now. It contains the
implementation of the useful (although currently defunct) heap-dump
feature.

With the removal of `boehm` and `go` GC support, the `hasSharedHeap`
constant is now always false, and is thus removed.

saem reviewed (approved):
The delete is strong int this one. :D

This gives renewed hope to improved gdb pretty printing support, as well.

saem commented:
/merge

saem commented:
/merge

(sorry, thought I saw something, my mistake)

zerbina committed:
assign: remove obsolete compiler runtime routines

zerbina committed:
assign: remove obsolete `objectInit`

It's part of the old runtime and no longer needed. Comments that reference
it in the compiler are also adjusted.

zerbina committed:
general `.compilerproc` cleanup

Remove obsolete `.compilerproc`s belonging to the old runtime.

zerbina committed:
mmdisp: clean up

* remove obsolete constants
* update documentation

zerbina committed:
clean up `malloc.nim`

Remove all the things related to the removed GCs.

zerbina committed:
gc_interface: remove obsolete prototypes

Neither the `GC_setStrategy` nor the `GC_ref`/`GC_unref` overload for
`seq`s and `string`s are implemented anywhere.

zerbina committed:
mmdisp: remove the `ForeignCell` API

It was neither used nor documented anywhere, and also does nothing when
using the destructor-based GCs.

zerbina committed:
remove `cellseqs_v1.nim`

It was only used by parts of the old runtime, and is thus obsolete.

zerbina committed:
alloc: remove dead code

* remove dead code related to the removed GCs
* port over some `alloc`/`dealloc` checks, as well as allocation tracker
  support

zerbina committed:
system: remove unnecessary `gcDestructors` usage

All modules only included when the C backend is used don't need to
check for `gcDestructors`, as there it is always defined.

zerbina committed:
threads: remove redundant `threadProcWrapStackFrame`

The procedure is only used a single time, and does nothing elses besides
dispatching to `threadProcWrapDispatch`.

zerbina committed:
doc: update `intern.rst`
